extensions-contrib/moving-average-query/pom.xml

```diff
@@ -0,0 +1,89 @@
+  ~ or more contributor license agreements.  See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership.  The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License.  You may obtain a copy of the License at
+  ~
+  ~   http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <modelVersion>4.0.0</modelVersion>
+
+  
+  <parent>
+    <artifactId>druid</artifactId>
+    <groupId>org.apache.druid</groupId>
+    <version>0.13.0-SNAPSHOT</version>
+    <relativePath>../../pom.xml</relativePath>
+  </parent>
+
+  <groupId>org.apache.druid.extensions.contrib</groupId>
+  <artifactId>druid-moving-average-query</artifactId>
+  <name>druid-moving-average-query</name>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.jmockit</groupId>
+      <artifactId>jmockit</artifactId>
+      <version>1.25</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.hamcrest</groupId>
+      <artifactId>hamcrest-library</artifactId>
+      <version>1.3</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.dataformat</groupId>
+      <artifactId>jackson-dataformat-yaml</artifactId>
+      <version>2.8.3</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.druid</groupId>
+      <artifactId>druid-api</artifactId>
+      <version>${project.parent.version}</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.druid</groupId>
+      <artifactId>druid-processing</artifactId>
+      <version>${project.parent.version}</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.druid</groupId>
+      <artifactId>druid-server</artifactId>
+      <version>${project.parent.version}</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.druid.extensions</groupId>
+      <artifactId>druid-datasketches</artifactId>
```

Comment:
This will not work unfortunately. Extensions cannot currently depend on other extensions. Modules of modules are not natively supported. Maybe this should be in the core datasketches?

Reply:
I'm not so sure about that. i.e. `materialized-view-selection` depends on `materialized-view-maintenance`. I tried running `pull deps` and got the dependency to be pulled along with the requested module:
```
java \
-cp "lib/*" \
-Ddruid.extensions.directory="extensions" \
-Ddruid.extensions.hadoopDependenciesDir="hadoop-dependencies" \
org.apache.druid.cli.Main tools pull-deps \
--no-default-hadoop \
-c org.apache.druid.extensions.contrib:materialized-view-selection:0.13.0-incubating

ls -1 extensions/materialized-view-selection/
materialized-view-maintenance-0.13.0-incubating.jar
materialized-view-selection-0.13.0-incubating.jar
```

I'm not sure how to test the distribution packaging + using `pull-deps` for unreleased code, as I haven't found documentation for it. perhaps you have some suggestion on how to do that?

Reply:
It will not automatically load the dependent extension, so docs should mention that the other extension must also be loaded manually to use this extension, but it can work.

The parquet extension does this too, depending on the avro extension to re-use stuff for the parquet-avro conversion parser https://github.com/apache/incubator-druid/blob/master/extensions-core/parquet-extensions/pom.xml#L44

Reply:
I think it depends on whether the dependency is `provided` or `compile`. IIRC `provided` means the user needs to be told to load both extensions and `compile` means the jar goes in both places. But I'm not sure how either of those mechanisms work if you need to use classes from the other extension.

Reply:
The ClassLoader needs to have the same class instance found to make them assignable. For Druid the application ClassLoader is the top one, and the extensions each have their own classloader. This means an extension can find classes in the application classloader but not other extensions. If something binds to a class found in the application classloader it is fine, but if it tries to pull in classes from another classloader in another extension, there is no way for it to resolve the class, so it should not work. I have no idea why the parquet one works if it isn't actually pulling in the classes in multiple places.

The exception to this rule is with Hadoop, where it puts all the junk on the same classloader, but that classloader is isolated from the application classloader. So maybe the parquet one only works in the hadoop classloader context.

THIS is why you get those weird logging errors sometimes with hadoop tasks. The application classloader binds JCL to a log4j impl, then the hadoop *ISOLATED* classloader loads up separate log4j classes (from the same jar, but different classloader) but is connected by a common system level classloader via the JCL. This confuses the heck out of the class matching system in log4j and you end up in a scenario where an impl can't assign to its interface because the impl is in one classloader but the interface is from another all isolated by how the JCL classes bind up vs how log4j classes bind up.

Reply:
I ran into this issue a lot with the original QTL extensions. If it does work for this extension I would love to know the mechanisms which allow it to do so and apply them elsewhere

Reply:
You'll notice the materialized view stuff actually copies the jars twice : https://github.com/apache/incubator-druid/blob/druid-0.13.0-incubating/extensions-contrib/materialized-view-selection/pom.xml#L41-L45 if you include both extensions. The reason is it is not listed as `provided`

Reply:
>The exception to this rule is with Hadoop, where it puts all the junk on the same classloader, but that classloader is isolated from the application classloader. So maybe the parquet one only works in the hadoop classloader context.

This is probably true since the parquet extension is only used for hadoop indexing, so the dependency only _appears_ to work, and likely means it will only work, as is, for hadoop.

Reply:
Thanks for these insights. So far I was able to use the module in queries both with and without the dataSketches dependency. My action item here is to document it better so people will know how to setup the module. I'll do that later when the review process is closer to finish.

Reply:
I have no idea how that works 
https://render.bitstrips.com/v2/cpanel/d8e66e64-ec9c-495f-947e-9bfe52d2523f-4c2a57dd-1478-47b6-825f-1037b33afc4d-v1.png?transparent=1&palette=1

But if it works it works!

Reply:
Actually, a recent change to datasketches (#6951) has broken this module due to DI errors. This makes it more clear for me how herarchial module dependency can become an issue. 
I've decided to remove datasketches support from this pr and add it back later, so it won't be a blocker.